### PR TITLE
1.21.0 Wind Charge Implementation

### PR DESCRIPTION
**Pull request by cjmustard**
## Introduction
Implementation for the 1.21.0 `Wind Charge`

### Relevant issues
N/A

## Details
- Non-iron doors and trapdoors are flipped.
- Fence gates are flipped.
- Buttons are pressed.
- Levers are flipped.
- Bells are rung and swung.
- Chorus flowers and pointed dripstone blocks break, dropping in item form.
- Lit candles (both standalone and on cake) are extinguished.

Wind charges can be deflected if hit with another projectile or a melee attack, similar to ghast fireballs.
Wind charges are distance attenuated by liquids such as water and lava. They also appear to fly in a straight line, similar to fireballs
Wind charges are set to despawn after five minuets due to the gravity less nature of the entity.

https://minecraft.wiki/w/Wind_Charge

## Changes
One method has been added to pocketmine\block\Block:
```php
	/**
	 * Called when this block is updated by a state altering projectile in the world.
	 *
	 * @param pocketmine\entity\projectile\Projectile $projectile The projectile affecting the block.
	 */
	public function onProjectileInteraction(pocketmine\entity\projectile\Projectile $projectile) : void {

	}
```

Both class' for the Wind Charge item and Projectile have been created:
- pocketmine\entity\projectile\WindCharge.php
- pocketmine\item\WindCharge.php

One class has been created to handle the wind charge explode particle:
- pocketmine\world\particle\WindExplosionParticle.php

One class has been added to handle wind charge burst sound:
- pocketmine/world/sound/WindChargeBurstSound.php

The Wind Charge has been implemented as a vanilla item.
The Wind Charge Entity has been implemented as an Entity.

### API changes
N/A

### Behavioural changes
N/A

## Backwards compatibility
N/A

## Follow-up
N/A

## Tests
I tested this PR by doing the following (tick all that apply):
- [ ] Writing PHPUnit tests (commit these in the `tests/phpunit` folder)
- [x] Playtesting using a Minecraft client (provide screenshots or a video)
- [ ] Writing a test plugin (provide the code and sample output)
- [ ] Other (provide details)

# Behavior With Other Blocks
https://github.com/pmmp/PocketMine-MP/assets/80650734/e27ca3d6-f086-432b-bff1-72be8c252557

# Behavior With Players (shows less effective under water.)
https://github.com/pmmp/PocketMine-MP/assets/80650734/2576fc31-f9cb-46de-a2c6-28e39964849a

# Behavior Multi Hit
https://github.com/pmmp/PocketMine-MP/assets/80650734/0250cdab-350e-4586-bf7a-2340195e0429

### CODE REVIEW

**Comment by pandaaaBE** [comment]
image

When I am standing still and aim to the ground.. In Vanilla i get boosted maybe 2.5 blocks and with this implementation nothing happens

(already fixed)

**Comment by pandaaaBE** [comment]
image

I can spam the wind charges to the sky and they will not despawn, this should be fixed somehow

(already fixed)

**Comment by pandaaaBE** [comment]
The sound "wind_charge.burst" is missing

(already added lol)

**Comment by cjmustard** [comment]
all issues mentioned previously should be fixed now!

File: src/entity/projectile/WindCharge.php [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+			$world->addSound(
+				$affectedBlock->getPosition()->add(0.5, 0.5, 0.5),
+				$affectedBlock->isActivated() ? new RedstonePowerOnSound() : new RedstonePowerOffSound()
+			);
+
+			return;
+		}
+
+		if($affectedBlock instanceof Bell) {
+			/** @var Bell $affectedBlock */
```
Comment (SOF3) [inline src/entity/projectile/WindCharge.php]: Is this redundant? Most static analysis tools can infer this from the `instanceof` line.

Comment (cjmustard) [reply src/entity/projectile/WindCharge.php]: yes very much so

File: src/entity/projectile/WindCharge.php [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		$maxX = (int) ceil($source->x + $radius + 1);
+		$minY = (int) floor($source->y - $radius - 1);
+		$maxY = (int) ceil($source->y + $radius + 1);
+		$minZ = (int) floor($source->z - $radius - 1);
+		$maxZ = (int) ceil($source->z + $radius + 1);
+
+		return new AxisAlignedBB($minX, $minY, $minZ, $maxX, $maxY, $maxZ);
+	}
+
+	private function checkAffect(Block $affectedBlock) : void {
```
Comment (Matze997) [inline src/entity/projectile/WindCharge.php]: More customizations should be possible here
Maybe event(s)?

Also method name could be missleading

File: src/entity/projectile/WindCharge.php [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+use pocketmine\world\sound\WindChargeBurstSound;
+
+use function ceil;
+use function floor;
+use function round;
+use function time;
+
+class WindCharge extends Throwable{
+	protected float $radius = 2.5;
+	protected float $damage = 0.5;
```
Comment (nicholass003) [inline src/entity/projectile/WindCharge.php]: damage value is 1 (half heart)

Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: These should be constants instead of variables

File: src/entity/projectile/WindCharge.php [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+	public static function getNetworkTypeId() : string{ return EntityIds::WIND_CHARGE_PROJECTILE; }
+
+	protected function getInitialDragMultiplier() : float{ return 0; }
+	protected function getInitialGravity() : float{ return 0; }
+
+	protected function initEntity(CompoundTag $nbt) : void{
+		parent::initEntity($nbt);
+
+		$this->createdAt = time();
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: All minecraft related timing uses ticks, you should be doing like this

Comment (IvanCraft623) [reply src/entity/projectile/WindCharge.php]: Also there's already a variable that tracks entity's lived ticks: https://github.com/pmmp/PocketMine-MP/blob/585aee9386a787c95e73dd0a05ffca8329606b68/src/entity/Entity.php#L137

File: src/entity/projectile/WindCharge.php [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	public function attack(EntityDamageEvent $source) : void{
+		if(!$source instanceof EntityDamageByEntityEvent) return;
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]:
```suggestion
		if(!$source instanceof EntityDamageByEntityEvent || ($entity = $source->getDamager()) === null){
				return;
		}
```

File: src/entity/projectile/WindCharge.php [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+		if(($entity = $source->getDamager()) == null) return;
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: If you do the suggestion above you can remove this line

File: src/entity/projectile/WindCharge.php [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+		$this->setOwningEntity($entity);
+
+		$this->setMotion($entity->getDirectionVector()->multiply(1.5));
+	}
+
+	protected function onHitEntity(Entity $entityHit, RayTraceResult $hitResult) : void{
+		if($this->getOwningEntity() === null)$ev = new EntityDamageByEntityEvent($this, $entityHit, EntityDamageEvent::CAUSE_PROJECTILE, $this->damage);
+		else $ev = new EntityDamageByChildEntityEvent($this->getOwningEntity(), $this, $entityHit, EntityDamageEvent::CAUSE_PROJECTILE, $this->damage);
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: Always use `if-else` whith brackets `{}`

File: src/entity/projectile/WindCharge.php [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+		return true;
+	}
+
+	protected function onHit(ProjectileHitEvent $event) : void{
+		$source = $this->getLocation();
+
+		$this->getWorld()->addSound($event->getRayTraceResult()->getHitVector(), new WindChargeBurstSound());
+		$this->getWorld()->addParticle($source, new WindExplosionParticle());
+
+		$list = $source->getWorld()->getNearbyEntities($this->getBound($source, $this->radius));
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: Don't create a variable if you are only going to use a value once

Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: Use `World::getCollidingEntities()` instead of `getNearbyEntities` since this behavior should check if the entity is colliderable. Also pass `$this` as the `entity` algument, so this way you can be sure that this entity is not on the list.

File: src/entity/projectile/WindCharge.php [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+		foreach($list as $entity){
+			if($entity == $this) continue;
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: If you do mentioned above you can remove this

File: src/entity/projectile/WindCharge.php [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+			$entityPos = $entity->getPosition();
+			$distance = $entityPos->distance($source) / $this->radius;
+			$motion = $entityPos->subtractVector($source)->normalize();
+
+			($entity->isUnderwater()) ? $exposure = 0.5 : $exposure = 1.5;
+			$impact = (1 - $distance) * $exposure;
+
+			if ($impact <= 0) continue;
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: Avoid using on single line if, on pocketmine we use brackets code-style

File: src/entity/projectile/WindCharge.php [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+		}
+	}
+
+	private function getBound(Vector3 $source, float $radius) : AxisAlignedBB {
+		$minX = (int) floor($source->x - $radius - 1);
+		$maxX = (int) ceil($source->x + $radius + 1);
+		$minY = (int) floor($source->y - $radius - 1);
+		$maxY = (int) ceil($source->y + $radius + 1);
+		$minZ = (int) floor($source->z - $radius - 1);
+		$maxZ = (int) ceil($source->z + $radius + 1);
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: Again, don't create variables if you are gonna use it once

File: src/entity/projectile/WindCharge.php (hunk 2 above)
Comment (IvanCraft623) [reply src/entity/projectile/WindCharge.php]: Agree about method name

Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: Would be more consistent create a new method on the Block class and each block will decide it's own behavour

**Review by IvanCraft623 — CHANGES_REQUESTED** [review]
Quick review about code quality, Haven't reviewd a vanilla parity behavour

File: src/entity/projectile/WindCharge.php (hunk 2 above)
Comment (cjmustard) [reply src/entity/projectile/WindCharge.php]: ProjectileLaunchEvent & ProjectileHitEvent already exist

**Comment by cjmustard** [comment]
this is a clean version.

File: src/entity/projectile/WindCharge.php (hunk 3 above)
Comment (IvanCraft623) [reply src/entity/projectile/WindCharge.php]: Don't close a conversation if you haven't resolved it.

File: src/entity/projectile/WindCharge.php (hunk 6 above)
Comment (IvanCraft623) [reply src/entity/projectile/WindCharge.php]: Same here

Comment (cjmustard) [reply src/entity/projectile/WindCharge.php]: done !

File: src/entity/projectile/WindCharge.php (hunk 3 above)
Comment (cjmustard) [reply src/entity/projectile/WindCharge.php]: sorry !

File: src/entity/projectile/WindCharge.php [hunk 12]
```diff
@@ -41,8 +41,6 @@
 use function round;
 
 class WindCharge extends Throwable{
-	protected float $radius = 2.5;
```
Comment (IvanCraft623) [inline src/entity/projectile/WindCharge.php]: I told you about making them constants, no remvonving them

This is a constant btw
https://github.com/pmmp/PocketMine-MP/blob/585aee9386a787c95e73dd0a05ffca8329606b68/src/entity/projectile/Arrow.php#L52

File: src/block/Button.php [hunk 13]
```diff
@@ -91,4 +92,16 @@ public function onNearbyBlockChange() : void{
 	private function canBeSupportedAt(Block $block, int $face) : bool{
 		return $block->getAdjacentSupportType(Facing::opposite($face))->hasCenterSupport();
 	}
+
+	public function onWindChargeInteraction(WindCharge $windCharge) : void{
+		if($this->getTypeId() == BlockTypeIds::IRON_DOOR) {
```
Comment (Matze997) [inline src/block/Button.php]: Why do you check for an iron door in the button class?

File: src/block/Door.php [hunk 14]
```diff
@@ -176,4 +177,15 @@ public function getAffectedBlocks() : array{
 	private function canBeSupportedAt(Block $block) : bool{
 		return $block->getAdjacentSupportType(Facing::DOWN)->hasEdgeSupport();
 	}
+
+	public function onWindChargeInteraction(WindCharge $windCharge) : void{
+		if($this->getTypeId() == BlockTypeIds::IRON_DOOR) {
```
Comment (Matze997) [inline src/block/Door.php]: if(!self instanceof WoodenDoor){
    return;
}

Would make a bit more sense here

Comment (cjmustard) [reply src/block/Door.php]: careless porting from a separate file, thank you for this catch!

Comment (cjmustard) [reply src/block/Door.php]: but that wouldnt work beacuse of copper doors!

File: src/block/Button.php (hunk 13 above)
Comment (AIPTU) [inline src/block/Button.php]: Shouldn't it using '===' instead?

**Comment by IvanCraft623** [comment]
@ShockedPlot7560 What do you think about using `Block::onProjectileHit()` instead of creating a new function only for specific projectile logic. Would have to change it to accept null in $hitResult parameter, so it would be BC Break

**Comment by pandaaaBE** [comment]
> @ShockedPlot7560 What do you think about using `Block::onProjectileHit()` instead of creating a new function only for specific projectile logic. Would have to change it to accept null in $hitResult parameter, so it would be BC Break

I think a BC Break for such an item would be unnecessary.. Maybe it's something for another PR. 
Someone could make a PR for it after this PR has been merged, otherwise you would have to wait until PM6 for the item because of the BC break

File: src/block/Block.php [hunk 15]
```diff
@@ -526,7 +536,7 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	/**
 	 * Called when this block is attacked (left-clicked) by a player attempting to start breaking it in survival.
 	 *
-	 * @return bool if an action took place, prevents starting to break the block if true.
+	 * @return bool if an action took place, prs starting to break the block if true.
```
Comment (ShockedPlot7560) [inline src/block/Block.php]: unrelated change

File: src/block/Door.php (hunk 14 above)
Comment (ShockedPlot7560) [inline src/block/Door.php]: use ===

File: src/block/Trapdoor.php [hunk 16]
```diff
@@ -91,4 +92,15 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
...
+
+	public function onWindChargeInteraction(WindCharge $windCharge) : void{
+		if($this->getTypeId() == BlockTypeIds::IRON_TRAPDOOR) {
+			return;
+		}
+
+		$this->open = !$this->open;
+		$world = $this->position->getWorld();
+		$world->setBlock($this->position, $this);
+		$world->addSound($this->position, new DoorSound());
```
Comment (ShockedPlot7560) [inline src/block/Trapdoor.php]: Need to be moved into a function to be used here and in onInteract

File: src/block/Lever.php [hunk 17]
```diff
@@ -101,6 +102,16 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
...
 
+	public function onWindChargeInteraction(WindCharge $windCharge) : void{
+		$this->activated = !$this->activated;
+		$world = $this->position->getWorld();
+		$world->setBlock($this->position, $this);
+		$world->addSound(
+			$this->position->add(0.5, 0.5, 0.5),
+			$this->isActivated() ? new RedstonePowerOnSound() : new RedstonePowerOffSound()
+		);
+	}
```
Comment (ShockedPlot7560) [inline src/block/Lever.php]: same here, moove the logic into one function

File: src/block/Door.php [hunk 18]
```diff
@@ -176,4 +177,15 @@ public function getAffectedBlocks() : array{
...
+			return;
+		}
+
+		$this->open = !$this->open;
+		$world = $this->position->getWorld();
+		$world->setBlock($this->position, $this);
+		$world->addSound($this->position, new DoorSound());
```
Comment (ShockedPlot7560) [inline src/block/Door.php]: need to used the same logic as interact into one single function to be sure to update the top door block

File: src/block/Button.php [hunk 19]
```diff
@@ -91,4 +92,16 @@ public function onNearbyBlockChange() : void{
...
+			return;
+		}
+
+		$this->pressed = true;
+		$world = $this->position->getWorld();
+		$world->setBlock($this->position, $this);
+		$world->scheduleDelayedBlockUpdate($this->position, $this->getActivationTime());
+		$world->addSound($this->position->add(0.5, 0.5, 0.5), new RedstonePowerOnSound());
```
Comment (ShockedPlot7560) [inline src/block/Button.php]: move it in a single function to use it in interact with the same logic

**Review by ShockedPlot7560 — CHANGES_REQUESTED** [review]

**Comment by ShockedPlot7560** [comment]
> @ShockedPlot7560 What do you think about using `Block::onProjectileHit()` instead of creating a new function only for specific projectile logic. Would have to change it to accept null in $hitResult parameter, so it would be BC Break

Thats not really a projectileHit, I don't think we need to change this function signature. It's more a projectile effect or smth like this.

File: composer.lock
Comment (ShockedPlot7560) [inline composer.lock:1]: no need to be updated

File: src/block/Button.php [hunk 20]
```diff
@@ -91,4 +89,19 @@ public function onNearbyBlockChange() : void{
...
 		return $block->getAdjacentSupportType(Facing::opposite($face))->hasCenterSupport();
 	}
+
+	public function onProjectileInteraction(Projectile $projectile) : void{
+		if($projectile instanceof WindCharge) {
+			$this->toggle();
+		}
+	}
+
+	public function toggle() : void {
```
Comment (ShockedPlot7560) [inline src/block/Button.php]: Thats not really a toggle, more a press or smth like this on the button.

File: src/block/Door.php [hunk 21]
```diff
@@ -176,4 +168,34 @@ public function getAffectedBlocks() : array{
...
+		if($projectile instanceof WindCharge) {
+			if($this->getTypeId() === BlockTypeIds::IRON_DOOR) {
+				return;
+			}
+
+			$this->open = !$this->open;
+
+			$world = $this->position->getWorld();
+			$world->setBlock($this->position, $this);
+			$world->addSound($this->position, new DoorSound());
```
Comment (ShockedPlot7560) [inline src/block/Door.php]: use toggle ?

File: src/block/Lever.php [hunk 22]
```diff
@@ -91,19 +93,28 @@ public function onNearbyBlockChange() : void{
...
+		if($projectile instanceof WindCharge) {
+			$this->toggle();
+		}
+	}
+
 	private function canBeSupportedAt(Block $block, int $face) : bool{
 		return $block->getAdjacentSupportType($face)->hasCenterSupport();
 	}
 
-	//TODO
```
Comment (ShockedPlot7560) [inline src/block/Lever.php]: redstone isnt implemented, so it still to do

**Comment by cjmustard** [comment]
@ShockedPlot7560 the issues you have mentioned have been fixed, thanks for pointing those out!

File: src/entity/projectile/WindCharge.php [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+use pocketmine\event\entity\EntityDamageByEntityEvent;
+use pocketmine\event\entity\EntityDamageEvent;
+use pocketmine\event\entity\ProjectileHitEvent;
+use pocketmine\math\AxisAlignedBB;
+use pocketmine\math\RayTraceResult;
+use pocketmine\math\Vector3;
+use pocketmine\network\mcpe\protocol\types\entity\EntityIds;
+use pocketmine\world\particle\WindExplosionParticle;
+use pocketmine\world\sound\WindChargeBurstSound;
+
```
Comment (nicholass003) [inline src/entity/projectile/WindCharge.php]: fix code style here

Comment (cjmustard) [reply src/entity/projectile/WindCharge.php]: what exactly am I changing here?

Comment (cjmustard) [reply src/entity/projectile/WindCharge.php]: the line thank you @pandaa-be

**Comment by ShockedPlot7560** [comment]
Why have you merged the stable into it ? Its now not up to date with minor-next and the diff reading is difficult with a lot of unrelated changes.

**Comment by cjmustard** [comment]
@ShockedPlot7560 sorry, I am new to the whole git ecosystem, have I fixed this issue or is there something else I would need to do?

**Comment by cjmustard** [comment]
bump

File: composer.lock
Comment (ShockedPlot7560) [inline composer.lock:1]: Unrelated changes

File: src/block/Door.php [hunk 24]
```diff
@@ -176,4 +168,29 @@ public function getAffectedBlocks() : array{
...
 		return $block->getAdjacentSupportType(Facing::DOWN)->hasEdgeSupport();
 	}
+
+	public function onProjectileInteraction(Projectile $projectile) : void{
+		if($projectile instanceof WindCharge) {
+			if($this->getTypeId() === BlockTypeIds::IRON_DOOR) {
+				return;
+			}
+
+			$this->toggle();
```
Comment (ShockedPlot7560) [inline src/block/Door.php:178]: Is the door open when the wind charge interacts with it?

File: src/block/Lever.php [hunk 25]
```diff
@@ -91,18 +93,29 @@ public function onNearbyBlockChange() : void{
...
+
+	public function onProjectileInteraction(Projectile $projectile) : void{
+		if($projectile instanceof WindCharge) {
+			$this->toggle();
+		}
+	}
+
+	private function canBeSupportedAt(Block $block, int $face) : bool{
+		return $block->getAdjacentSupportType($face)->hasCenterSupport();
+	}
```
Comment (ShockedPlot7560) [inline src/block/Lever.php]: unrelated changes

File: src/block/Lever.php [hunk 26]
```diff
@@ -91,18 +93,29 @@ public function onNearbyBlockChange() : void{
...
+
+	public function toggle() : void {
 		$this->activated = !$this->activated;
 		$world = $this->position->getWorld();
 		$world->setBlock($this->position, $this);
 		$world->addSound(
 			$this->position->add(0.5, 0.5, 0.5),
-			$this->activated ? new RedstonePowerOnSound() : new RedstonePowerOffSound()
+			$this->isActivated() ? new RedstonePowerOnSound() : new RedstonePowerOffSound()
```
Comment (ShockedPlot7560) [inline src/block/Lever.php]: Direct access to the property is enough

File: composer.lock
Comment (cjmustard) [reply composer.lock:1]: I am not sure how this became out of sync, and can not seem to make it match. could you please help me out

File: src/block/Door.php (hunk 24 above)
Comment (cjmustard) [reply src/block/Door.php:178]: it should toggle the door either way it is facing

**Comment by cjmustard** [comment]
Closing to rebase at composer.lock level
